crates/utils/src/rate_limit/rate_limiter.rs

```diff
@@ -56,179 +75,227 @@ pub(crate) enum RateLimitType {
+
+    #[allow(clippy::indexing_slicing)]
+    let total_passes = group.check_total(action_type, now, adjusted_configs[action_type]);
+
+    let children_pass = group.children.check(
+      action_type,
+      now,
+      configs,
+      child_capacity_factors,
+      child_addr_parts,
+    );
+
+    total_passes && children_pass
   }
 
-  fn check_total(
+  fn remove_full_buckets(
     &mut self,
-    type_: RateLimitType,
     now: InstantSecs,
-    capacity: i32,
-    secs_to_refill: i32,
+    configs: EnumMap<ActionType, BucketConfig>,
   ) -> bool {
-    let capacity = capacity as f32;
-    let secs_to_refill = secs_to_refill as f32;
+    self.retain(|_key, group| {
+      let some_children_remaining = group.children.remove_full_buckets(now, configs);
+
+      // Evaluated if `some_children_remaining` is false
+      let total_has_refill_in_future = || {
+        group.total.into_iter().all(|(action_type, bucket)| {
+          #[allow(clippy::indexing_slicing)]
+          let config = configs[action_type];
+          bucket.update(now, config).tokens != config.capacity
+        })
+      };
 
-    #[allow(clippy::indexing_slicing)] // `EnumMap` has no `get` funciton
-    let bucket = &mut self.total[type_];
+      some_children_remaining || total_has_refill_in_future()
+    });
 
-    if bucket.tokens == UNINITIALIZED_TOKEN_AMOUNT {
-      bucket.tokens = capacity;
-    }
+    self.shrink_to_fit();
+
+    !self.is_empty()
+  }
+}
 
-    let secs_since_last_checked = now.secs_since(bucket.last_checked) as f32;
-    bucket.last_checked = now;
+impl MapLevel for () {
+  type CapacityFactors = ();
+  type AddrParts = ();
 
-    // For `secs_since_last_checked` seconds, increase `bucket.tokens`
-    // by `capacity` every `secs_to_refill` seconds
-    bucket.tokens += {
-      let tokens_per_sec = capacity / secs_to_refill;
-      secs_since_last_checked * tokens_per_sec
-    };
+  fn check(
+    &mut self,
+    _: ActionType,
+    _: InstantSecs,
+    _: EnumMap<ActionType, BucketConfig>,
+    _: Self::CapacityFactors,
+    _: Self::AddrParts,
+  ) -> bool {
+    true
+  }
+
+  fn remove_full_buckets(&mut self, _: InstantSecs, _: EnumMap<ActionType, BucketConfig>) -> bool {
+    false
+  }
+}
 
-    // Prevent `bucket.tokens` from exceeding `capacity`
-    if bucket.tokens > capacity {
-      bucket.tokens = capacity;
+impl<C: Default> RateLimitedGroup<C> {
+  fn new(now: InstantSecs, configs: EnumMap<ActionType, BucketConfig>) -> Self {
+    RateLimitedGroup {
+      total: configs.map(|_, config| Bucket {
+        last_checked: now,
+        tokens: config.capacity,
+      }),
+      // `HashMap::new()` or `()`
+      children: Default::default(),
     }
+  }
+
+  fn check_total(
+    &mut self,
+    action_type: ActionType,
+    now: InstantSecs,
+    config: BucketConfig,
+  ) -> bool {
+    #[allow(clippy::indexing_slicing)] // `EnumMap` has no `get` funciton
+    let bucket = &mut self.total[action_type];
```

Comment:
This might panic if its missing right? With a map get its always an `Option`.

Reply:
No, `EnumMap` does not allow missing values